mqtt/src/main/scala/akka/stream/alpakka/mqtt/impl/MqttFlowStage.scala

```diff
@@ -44,249 +44,288 @@ private[mqtt] final class MqttFlowStage(connectionSettings: MqttConnectionSettin
-
-      private val onSubscribe: AsyncCallback[Try[IMqttToken]] = getAsyncCallback[Try[IMqttToken]] { conn =>
-        subscriptionPromise.complete(conn.map(_ => {
-          log.debug("subscription established")
-          Done
-        }))
-        pull(in)
-      }
+    val logic = new MqttFlowStageLogic[MqttMessage](in,
+                                                    out,
+                                                    shape,
+                                                    subscriptionPromise,
+                                                    connectionSettings,
+                                                    subscriptions,
+                                                    bufferSize,
+                                                    defaultQoS,
+                                                    manualAcks)
+    (logic, subscriptionPromise.future)
+  }
+}
 
-      private val onConnect: AsyncCallback[IMqttAsyncClient] =
-        getAsyncCallback[IMqttAsyncClient]((client: IMqttAsyncClient) => {
-          log.debug("connected")
-          if (subscriptions.nonEmpty) {
-            if (manualAcks) client.setManualAcks(true)
-            val (topics, qoses) = subscriptions.unzip
-            client.subscribe(topics.toArray, qoses.map(_.value).toArray, (), asActionListener(onSubscribe.invoke))
-          } else {
-            subscriptionPromise.complete(SuccessfullyDone)
-            pull(in)
-          }
-        })
-
-      private val onConnectionLost: AsyncCallback[Throwable] = getAsyncCallback[Throwable](failStageWith)
-
-      private val onMessageAsyncCallback: AsyncCallback[MqttMessageWithAck] =
-        getAsyncCallback[MqttMessageWithAck] { message =>
-          if (isAvailable(out)) {
-            pushDownstream(message)
-          } else if (queue.size + 1 > bufferSize) {
-            failStageWith(new RuntimeException(s"Reached maximum buffer size $bufferSize"))
-          } else {
-            queue.enqueue(message)
-          }
-        }
+class MqttFlowStageLogic[I](in: Inlet[I],
+                            out: Outlet[MqttMessageWithAck],
+                            shape: Shape,
+                            subscriptionPromise: Promise[Done],
+                            connectionSettings: MqttConnectionSettings,
+                            subscriptions: Map[String, MqttQoS],
+                            bufferSize: Int,
+                            defaultQoS: MqttQoS,
+                            manualAcks: Boolean)
+    extends GraphStageLogic(shape)
+    with StageLogging
+    with InHandler
+    with OutHandler {
+
+  import akka.stream.alpakka.mqtt.impl.MqttFlowStageLogic._
+
+  private val backpressurePahoClient = new Semaphore(bufferSize)
+  private var pendingMsg = Option.empty[I]
+  private val queue = mutable.Queue[MqttMessageWithAck]()
+  private val unackedMessages = new AtomicInteger()
+
+  protected def handleDeliveryComplete(token: IMqttDeliveryToken): Unit = ()
```

Comment:
extracted method for easy override

Reply:
the original method is inside the callback.
i prefer extract this method instead to override the entire callback